dedupe code in all.py and single.py

anthonyduong9 opened the pull request:
### Problem

- `all.py` and `single.py` both contain functions `_get_safe_dtype()` and `_safe_cast()`
- `all.py` and `single.py` duplicate `get_layer_num_from_sae_id()`
- A large block of code to calculate DFA values has been practically duplicated in `all.py` and `single.py`, and so it seems unlikely we'll want to change it in two places.

### Fix 

- Moving one pair of functions `_get_safe_dtype()` and `_safe_cast()` to `shared.py` and deleting the other pair.
- Moving `get_layer_num_from_sae_id()` to `shared.py`.
- Extracting the large block of code to `calculate_per_source_dfa()`.

### Testing

- I compared code I thought to be duplicated at www.diffchecker.com.
- I ran `make check-ci`.

anthonyduong9 committed:
dedupe code in all.py and single.py

hijohnnylin commented:
@anthonyduong9 thanks for the update, sorry for slowness on getting around to it!

we need test(s) that the before/after results are the same before merging this - can you write them, or let me know what info you'd need to write them? if you don't have time for it lmk - i can work on it.

anthonyduong9 commented:
@hijohnnylin I've added tests for `calculate_dfa()` and `ActivationProcessor.process_activations()`, since if you start from the functions where we extracted `calculate_per_source_dfa()` and go to where they're called, and so on, those're the public functions you end up at. They pass on `main` (save for one that references `calculate_per_source_dfa()`, since it doesn't exist), as well as this branch. The tests for `calculate_dfa()` largely match [this](https://github.com/jbloomAus/SAEDashboard/blob/9bef8178b7a5043a5849aff9a10503994a03eca5/tests/unit/test_dfa_calculator.py) in SAEDashboard, so assuming it's correct, we can have more confidence.
